src/main/java/htsjdk/samtools/cram/encoding/CramCodec.java

```diff
@@ -0,0 +1,29 @@
+package htsjdk.samtools.cram.encoding;
+
+/**
+ * An interface that defines requirements for serializing/deserializing objects into and from a stream.
+ *
+ * @param <T> data series type to be read or written
+ */
+public interface CramCodec<T> {
+    /**
+     * Read a single object from the stream
+     *
+     * @return an object from the stream
+     */
+    T read();
+
+    /**
+     * Read a array of specified length from the stream
+     *
+     * @param length the number of elements to read
+     * @return an object from the stream
+     */
+    T read(final int length);
```

Comment:
On further investigation, this can be removed!

It was only used by the External Byte Array encoding, which is not actually valid according to the spec, and seems to only exist due to mis-typing of the sub-encodings of the Byte Array Len encoding.

Reply:
I don't like this either but I haven't found a good solution yet.  I have a partial attempt in a local branch which I may come back to later.